File: ai_runtime/model_event.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Mapping

MODEL_EVENT_SCHEMA_VERSION = "1.0"


@dataclass(frozen=True)
class BoundingBoxV1:
    x: float
    y: float
    width: float
    height: float

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "BoundingBoxV1":
        box = cls(
            x=_finite_number(value.get("x"), "geometry.bounding_box.x"),
            y=_finite_number(value.get("y"), "geometry.bounding_box.y"),
            width=_finite_number(value.get("width"), "geometry.bounding_box.width"),
            height=_finite_number(value.get("height"), "geometry.bounding_box.height"),
        )
        if box.width < 0 or box.height < 0:
            raise ValueError("geometry bounding-box width/height must be non-negative")
        return box

    def to_dict(self) -> dict[str, float]:
        return {"x": self.x, "y": self.y, "width": self.width, "height": self.height}


@dataclass(frozen=True)
class KeypointV1:
    x: float
    y: float
    confidence: float | None = None
    label: str | None = None

    @classmethod
    def from_dict(cls, value: Mapping[str, Any], index: int) -> "KeypointV1":
        confidence = _optional_unit_interval(
            value.get("confidence"), f"geometry.keypoints[{index}].confidence"
        )
        label = _optional_text(value.get("label"), f"geometry.keypoints[{index}].label")
        return cls(
            x=_finite_number(value.get("x"), f"geometry.keypoints[{index}].x"),
            y=_finite_number(value.get("y"), f"geometry.keypoints[{index}].y"),
            confidence=confidence,
            label=label,
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "x": self.x,
            "y": self.y,
            "confidence": self.confidence,
            "label": self.label,
        }


@dataclass(frozen=True)
class ModelGeometryV1:
    coordinate_space: str | None = None
    bounding_box: BoundingBoxV1 | None = None
    keypoints: tuple[KeypointV1, ...] = ()
    vector: tuple[float, ...] | None = None
    region_id: str | None = None

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "ModelGeometryV1":
        coordinate_space = _optional_text(
            value.get("coordinate_space"), "geometry.coordinate_space"
        )
        region_id = _optional_text(value.get("region_id"), "geometry.region_id")

        box_value = value.get("bounding_box")
        if box_value is not None and not isinstance(box_value, Mapping):
            raise ValueError("geometry.bounding_box must be an object or null")
        bounding_box = BoundingBoxV1.from_dict(box_value) if box_value is not None else None

        keypoint_values = value.get("keypoints", [])
        if not isinstance(keypoint_values, list):
            raise ValueError("geometry.keypoints must be an array")
        keypoints: list[KeypointV1] = []
        for index, item in enumerate(keypoint_values):
            if not isinstance(item, Mapping):
                raise ValueError(f"geometry.keypoints[{index}] must be an object")
            keypoints.append(KeypointV1.from_dict(item, index))

        vector_value = value.get("vector")
        vector: tuple[float, ...] | None = None
        if vector_value is not None:
            if not isinstance(vector_value, list):
                raise ValueError("geometry.vector must be an array or null")
            vector = tuple(
                _finite_number(item, f"geometry.vector[{index}]")
                for index, item in enumerate(vector_value)
            )

        return cls(
            coordinate_space=coordinate_space,
            bounding_box=bounding_box,
            keypoints=tuple(keypoints),
            vector=vector,
            region_id=region_id,
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "coordinate_space": self.coordinate_space,
            "bounding_box": self.bounding_box.to_dict() if self.bounding_box else None,
            "keypoints": [keypoint.to_dict() for keypoint in self.keypoints],
            "vector": list(self.vector) if self.vector is not None else None,
            "region_id": self.region_id,
        }


@dataclass(frozen=True)
class ValidityIntervalV1:
    start_timestamp_ns: int
    end_timestamp_ns: int | None = None

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "ValidityIntervalV1":
        start = _non_negative_int(value.get("start_timestamp_ns"), "validity_interval.start_timestamp_ns")
        end_value = value.get("end_timestamp_ns")
        end = (
            _non_negative_int(end_value, "validity_interval.end_timestamp_ns")
            if end_value is not None
            else None
        )
        interval = cls(start_timestamp_ns=start, end_timestamp_ns=end)
        interval.validate()
        return interval

    def validate(self) -> None:
        if self.end_timestamp_ns is not None and self.end_timestamp_ns < self.start_timestamp_ns:
            raise ValueError(
                "validity_interval.end_timestamp_ns must be greater than or equal to start_timestamp_ns"
            )

    def to_dict(self) -> dict[str, int | None]:
        return {
            "start_timestamp_ns": self.start_timestamp_ns,
            "end_timestamp_ns": self.end_timestamp_ns,
        }


@dataclass(frozen=True)
class ModelEventV1:
    schema_version: str
    session_id: str
    event_id: str
    source_frame_id: int | None
    capture_timestamp_ns: int
    inference_timestamp_ns: int
    model_id: str
    model_version: str
    track_id: str | None
    class_id: str
    confidence: float | None
    quality: float | None
    geometry: ModelGeometryV1 | None
    validity_interval: ValidityIntervalV1
    metadata: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "ModelEventV1":
        _require_core_fields(value)

        source_frame_value = value.get("source_frame_id")
        source_frame_id = (
            _non_negative_int(source_frame_value, "source_frame_id")
            if source_frame_value is not None
            else None
        )

        geometry_value = value.get("geometry")
        if geometry_value is not None and not isinstance(geometry_value, Mapping):
            raise ValueError("geometry must be an object or null")
        geometry = ModelGeometryV1.from_dict(geometry_value) if geometry_value is not None else None

        validity_value = value.get("validity_interval")
        if not isinstance(validity_value, Mapping):
            raise ValueError("validity_interval must be an object")

        metadata_value = value.get("metadata")
        if not isinstance(metadata_value, dict):
            raise ValueError("metadata must be an object")

        event = cls(
            schema_version=_required_text(value.get("schema_version"), "schema_version"),
            session_id=_required_text(value.get("session_id"), "session_id"),
            event_id=_required_text(value.get("event_id"), "event_id"),
            source_frame_id=source_frame_id,
            capture_timestamp_ns=_non_negative_int(
                value.get("capture_timestamp_ns"), "capture_timestamp_ns"
            ),
            inference_timestamp_ns=_non_negative_int(
                value.get("inference_timestamp_ns"), "inference_timestamp_ns"
            ),
            model_id=_required_text(value.get("model_id"), "model_id"),
            model_version=_required_text(value.get("model_version"), "model_version"),
            track_id=_optional_text(value.get("track_id"), "track_id"),
            class_id=_required_text(value.get("class_id"), "class_id"),
            confidence=_optional_unit_interval(value.get("confidence"), "confidence"),
            quality=_optional_unit_interval(value.get("quality"), "quality"),
            geometry=geometry,
            validity_interval=ValidityIntervalV1.from_dict(validity_value),
            metadata=dict(metadata_value),
        )
        event.validate()
        return event

    def validate(self) -> None:
        if self.schema_version != MODEL_EVENT_SCHEMA_VERSION:
            raise ValueError(f"schema_version must be {MODEL_EVENT_SCHEMA_VERSION}")
        _required_text(self.session_id, "session_id")
        _required_text(self.event_id, "event_id")
        _required_text(self.model_id, "model_id")
        _required_text(self.model_version, "model_version")
        _required_text(self.class_id, "class_id")
        if self.track_id is not None:
            _required_text(self.track_id, "track_id")
        _non_negative_int(self.capture_timestamp_ns, "capture_timestamp_ns")
        _non_negative_int(self.inference_timestamp_ns, "inference_timestamp_ns")
        if self.inference_timestamp_ns < self.capture_timestamp_ns:
            raise ValueError(
                "inference_timestamp_ns must be greater than or equal to capture_timestamp_ns"
            )
        _optional_unit_interval(self.confidence, "confidence")
        _optional_unit_interval(self.quality, "quality")
        self.validity_interval.validate()

    def to_dict(self) -> dict[str, Any]:
        self.validate()
        return {
            "schema_version": self.schema_version,
            "session_id": self.session_id,
            "event_id": self.event_id,
            "source_frame_id": self.source_frame_id,
            "capture_timestamp_ns": self.capture_timestamp_ns,
            "inference_timestamp_ns": self.inference_timestamp_ns,
            "model_id": self.model_id,
            "model_version": self.model_version,
            "track_id": self.track_id,
            "class_id": self.class_id,
            "confidence": self.confidence,
            "quality": self.quality,
            "geometry": self.geometry.to_dict() if self.geometry else None,
            "validity_interval": self.validity_interval.to_dict(),
            "metadata": dict(self.metadata),
        }


_CORE_FIELDS = {
    "schema_version",
    "session_id",
    "event_id",
    "source_frame_id",
    "capture_timestamp_ns",
    "inference_timestamp_ns",
    "model_id",
    "model_version",
    "track_id",
    "class_id",
    "confidence",
    "quality",
    "geometry",
    "validity_interval",
    "metadata",
}


def _require_core_fields(value: Mapping[str, Any]) -> None:
    missing = sorted(_CORE_FIELDS - set(value.keys()))
    if missing:
        raise ValueError("missing ModelEventV1 core fields: " + ", ".join(missing))


def _required_text(value: Any, field_name: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{field_name} must be a non-empty string")
    return value.strip()


def _optional_text(value: Any, field_name: str) -> str | None:
    if value is None:
        return None
    return _required_text(value, field_name)


def _non_negative_int(value: Any, field_name: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{field_name} must be an integer")
    if value < 0:
        raise ValueError(f"{field_name} must be non-negative")
    return value


def _finite_number(value: Any, field_name: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"{field_name} must be a number")
    number = float(value)
    if number != number or number in (float("inf"), float("-inf")):
        raise ValueError(f"{field_name} must be finite")
    return number


def _optional_unit_interval(value: Any, field_name: str) -> float | None:
    if value is None:
        return None
    number = _finite_number(value, field_name)
    if not 0.0 <= number <= 1.0:
        raise ValueError(f"{field_name} must be between 0 and 1")
    return number

File: native/brain_core/src/api/model_event.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const MODEL_EVENT_SCHEMA_VERSION: &str = "1.0";
pub const MODEL_EVENT_CORE_FIELDS: [&str; 15] = [
    "schema_version",
    "session_id",
    "event_id",
    "source_frame_id",
    "capture_timestamp_ns",
    "inference_timestamp_ns",
    "model_id",
    "model_version",
    "track_id",
    "class_id",
    "confidence",
    "quality",
    "geometry",
    "validity_interval",
    "metadata",
];

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct BoundingBoxV1 {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct KeypointV1 {
    pub x: f32,
    pub y: f32,
    #[serde(default)]
    pub confidence: Option<f32>,
    #[serde(default)]
    pub label: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct ModelGeometryV1 {
    #[serde(default)]
    pub coordinate_space: Option<String>,
    #[serde(default)]
    pub bounding_box: Option<BoundingBoxV1>,
    #[serde(default)]
    pub keypoints: Vec<KeypointV1>,
    #[serde(default)]
    pub vector: Option<Vec<f32>>,
    #[serde(default)]
    pub region_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ValidityIntervalV1 {
    pub start_timestamp_ns: u64,
    #[serde(default)]
    pub end_timestamp_ns: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ModelEventV1 {
    pub schema_version: String,
    pub session_id: String,
    pub event_id: String,
    #[serde(default)]
    pub source_frame_id: Option<u64>,
    pub capture_timestamp_ns: u64,
    pub inference_timestamp_ns: u64,
    pub model_id: String,
    pub model_version: String,
    #[serde(default)]
    pub track_id: Option<String>,
    pub class_id: String,
    #[serde(default)]
    pub confidence: Option<f32>,
    #[serde(default)]
    pub quality: Option<f32>,
    #[serde(default)]
    pub geometry: Option<ModelGeometryV1>,
    pub validity_interval: ValidityIntervalV1,
    #[serde(default)]
    pub metadata: BTreeMap<String, Value>,
}

impl ModelEventV1 {
    pub fn validate(&self) -> Result<(), String> {
        if self.schema_version != MODEL_EVENT_SCHEMA_VERSION {
            return Err(format!(
                "schema_version must be {MODEL_EVENT_SCHEMA_VERSION}"
            ));
        }
        require_text(&self.session_id, "session_id")?;
        require_text(&self.event_id, "event_id")?;
        require_text(&self.model_id, "model_id")?;
        require_text(&self.model_version, "model_version")?;
        require_text(&self.class_id, "class_id")?;
        if let Some(track_id) = &self.track_id {
            require_text(track_id, "track_id")?;
        }
        if self.inference_timestamp_ns < self.capture_timestamp_ns {
            return Err(
                "inference_timestamp_ns must be greater than or equal to capture_timestamp_ns"
                    .into(),
            );
        }
        validate_unit_interval(self.confidence, "confidence")?;
        validate_unit_interval(self.quality, "quality")?;
        self.validity_interval.validate()?;
        if let Some(geometry) = &self.geometry {
            geometry.validate()?;
        }
        Ok(())
    }

    pub fn to_json(&self) -> Result<String, String> {
        self.validate()?;
        serde_json::to_string(self).map_err(|error| error.to_string())
    }

    pub fn from_json(value: &str) -> Result<Self, String> {
        let raw: Value = serde_json::from_str(value).map_err(|error| error.to_string())?;
        let object = raw
            .as_object()
            .ok_or_else(|| "ModelEventV1 must be a JSON object".to_string())?;
        let missing: Vec<&str> = MODEL_EVENT_CORE_FIELDS
            .iter()
            .copied()
            .filter(|field| !object.contains_key(*field))
            .collect();
        if !missing.is_empty() {
            return Err(format!(
                "missing ModelEventV1 core fields: {}",
                missing.join(", ")
            ));
        }
        let event: Self = serde_json::from_value(raw).map_err(|error| error.to_string())?;
        event.validate()?;
        Ok(event)
    }
}

impl ValidityIntervalV1 {
    pub fn validate(&self) -> Result<(), String> {
        if let Some(end_timestamp_ns) = self.end_timestamp_ns {
            if end_timestamp_ns < self.start_timestamp_ns {
                return Err(
                    "validity_interval.end_timestamp_ns must be greater than or equal to start_timestamp_ns"
                        .into(),
                );
            }
        }
        Ok(())
    }
}

impl ModelGeometryV1 {
    pub fn validate(&self) -> Result<(), String> {
        if let Some(coordinate_space) = &self.coordinate_space {
            require_text(coordinate_space, "geometry.coordinate_space")?;
        }
        if let Some(region_id) = &self.region_id {
            require_text(region_id, "geometry.region_id")?;
        }
        if let Some(bounding_box) = &self.bounding_box {
            validate_finite(bounding_box.x, "geometry.bounding_box.x")?;
            validate_finite(bounding_box.y, "geometry.bounding_box.y")?;
            validate_finite(bounding_box.width, "geometry.bounding_box.width")?;
            validate_finite(bounding_box.height, "geometry.bounding_box.height")?;
            if bounding_box.width < 0.0 || bounding_box.height < 0.0 {
                return Err("geometry bounding-box width/height must be non-negative".into());
            }
        }
        for (index, keypoint) in self.keypoints.iter().enumerate() {
            validate_finite(keypoint.x, &format!("geometry.keypoints[{index}].x"))?;
            validate_finite(keypoint.y, &format!("geometry.keypoints[{index}].y"))?;
            validate_unit_interval(
                keypoint.confidence,
                &format!("geometry.keypoints[{index}].confidence"),
            )?;
            if let Some(label) = &keypoint.label {
                require_text(label, &format!("geometry.keypoints[{index}].label"))?;
            }
        }
        if let Some(vector) = &self.vector {
            for (index, value) in vector.iter().enumerate() {
                validate_finite(*value, &format!("geometry.vector[{index}]"))?;
            }
        }
        Ok(())
    }
}

fn require_text(value: &str, field: &str) -> Result<(), String> {
    if value.trim().is_empty() {
        return Err(format!("{field} must be a non-empty string"));
    }
    Ok(())
}

fn validate_unit_interval(value: Option<f32>, field: &str) -> Result<(), String> {
    if let Some(value) = value {
        validate_finite(value, field)?;
        if !(0.0..=1.0).contains(&value) {
            return Err(format!("{field} must be between 0 and 1"));
        }
    }
    Ok(())
}

fn validate_finite(value: f32, field: &str) -> Result<(), String> {
    if !value.is_finite() {
        return Err(format!("{field} must be finite"));
    }
    Ok(())
}
